### PR TITLE
Wasm

### PR DESCRIPTION
**Pull request by zacharycarter**

### CODE REVIEW

**Comment by define-private-public** [comment]
Do you want me to do a review of this?

**Comment by zacharycarter** [comment]
That would be great if you wouldn't mind.

**Comment by define-private-public** [comment]
Sure, I can look through it when I get home.  Can you assign it to me?  I haven't played around with Web Assembly yet.  Where can I get started so I can run the stuff.

Also, do you think it we be possible to give me some issue tracker privileges?  All I can do at the moment is open, comment, and close issues.  I'd like to be able to do things such as tag tickets, set milestones, and other things.

**Comment by zacharycarter** [comment]
Sure I can do that - do you think you could hop into gitter / IRC tonight? I want to discuss a few things with you regarding the timer module too.

I think we should drop the TImer object and just have the module expose the functions that are there and move the vars under the timer object to globals. I don't see much point in having a timer object and I think it's going to be problematic.

I'm writing network code now where I want to be able to access the engine's time globally in a callback method and I can't do it. Do you see any benefits to having an object? I think if it's modular like the rest of the engine, it's more beneficial and more easily accessible.

Sorry for the hijack.

**Comment by define-private-public** [comment]
Can you put the thing about the Timer module in a separate ticket?

File: examples/data/shaders/glsl100/animation/forward.fs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#version 100
```
Comment (define-private-public) [inline examples/data/shaders/glsl100/animation/forward.fs:1]: IIRC, we're targeting GLSL 330.

File: examples/data/shaders/glsl100/animation/forward.vs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#version 100
```
Comment (define-private-public) [inline examples/data/shaders/glsl100/animation/forward.vs:1]: IIRC, we're targeting GLSL 330.

File: examples/src/zengine_examples/00_Minimal/main.nim [hunk 3]
```diff
@@ -1,98 +1,204 @@
 # Minimal example to open up a zengine window
 
-import zengine, sdl2, opengl, glm
+import zengine, opengl, glm, jsbind.emscripten, times, sdl2
```
Comment (define-private-public) [inline examples/src/zengine_examples/00_Minimal/main.nim]: What does jsbind do when we're not working with a JS target?  I don't know how the module works.  Can you give me some documentation?

File: examples/src/zengine_examples/00_Minimal/main.nim [hunk 4]
```diff
@@ -1,98 +1,204 @@
...
+  gcRequested = true
+
+proc startApplication() =
+  zengine.init(ScreenWidth, ScreenHeight, "Zengine example: 00_Minimal")
+  start()
+  camera.setMode(CameraMode.Free)
+  shader = loadShader("examples/data/shaders/glsl100/animation/forward.vs", "examples/data/shaders/glsl100/animation/forward.fs")
+  model = loadModel("examples/data/models/mutant/mutant_idle.dae", shader)
+  
+proc runGC() =
```
Comment (define-private-public) [inline examples/src/zengine_examples/00_Minimal/main.nim]: This looks new.  Can you explain it a little more why we're manually taking control of the GC?

File: examples/src/zengine_examples/00_Minimal/main.nim [hunk 5]
```diff
@@ -1,98 +1,204 @@
...
+template runApplication*(initCode: typed): stmt =
+  initFunc = proc() =
+      initCode
+  emscripten_set_main_loop(mainLoopPreload, 0, 1)
+
+runApplication:
+  startApplication()
+  
+
+# # State variables
```
Comment (define-private-public) [inline examples/src/zengine_examples/00_Minimal/main.nim]: All this comment stuff looks like some leftover cruft.  Is that the case?  If so, then purge it.

File: src/zengine/zgl.nim [hunk 6]
```diff
@@ -324,34 +348,73 @@ proc loadDefaultShader(): Shader =
     }
   """
 
-  let fDefaultShaderStr = """
-    #version 330
-    in vec2 exFragTexCoord;
-    in vec4 exColor;
-    out vec4 outColor;
+  fDefaultShaderStr = """
+    #version 100
```
Comment (define-private-public) [inline src/zengine/zgl.nim:352]: Why do we need to go back to GLSL 100?

**Review by define-private-public — CHANGES_REQUESTED** [review]
Many RFCs inline.  But some general things:

1. Let's keep the `Timer` module thing to it's own thing, unless it's absolutely required for this change.
2. I'm a little iffy about the use of `when defined(emscripten)` everywhere.  While it's after than C's #ifdef hell, there's got to be something better.  At least for ZGL, do you think it might be possible to make a `zgl_emscripten.nim` file?  We should figure out how much code duplication that would be first...

File: examples/data/shaders/glsl100/animation/forward.vs (hunk 2 above)
Comment (zacharycarter) [reply examples/data/shaders/glsl100/animation/forward.vs:1]: We can't for the browser - we have to target GLSL 100 for webgl

File: examples/src/zengine_examples/00_Minimal/main.nim (hunk 4 above)
Comment (zacharycarter) [reply examples/src/zengine_examples/00_Minimal/main.nim]: We're only doing this for emscripten

File: src/zengine/zgl.nim (hunk 6 above)
Comment (zacharycarter) [reply src/zengine/zgl.nim:352]: Only for emscripten again

**Comment by define-private-public** [comment]
Okay, I haven't looked too much into emscripten and GLSL stuff, but it looks like the best supported version is actually OpenGL **ES** 2.0/3.0

http://kripken.github.io/emscripten-site/docs/porting/multimedia_and_graphics/OpenGL-support.html

Maybe instead of using Desktop GLSL, we should be using the ES version.  That would also give us a boost when it comes to try running zengine on mobile.  Many desktop graphics cards support ES anyways.  It's just more portable.

**Comment by define-private-public** [comment]
I'll take another look at this later.  Are we set on having one global Timer object?

**Comment by zacharycarter** [comment]
Yeah we are - I added the changes into this PR.

It supports the WebGL friendly subset of OpenGL ES 2 - meaning we can get away with using the OpenGL bindings. Regardless, webgl 1 requires 100 level shaders.

**Review by define-private-public — APPROVED** [review]
I think this should suffice for now.  While I'm not the biggest fan of all this `when defined emscripten` that's inline, but it's not explosively huge at the moment.

There has got to be a better way to do this.  Personally, maybe having an `zgl_emscripten.nim` file that these changes are broken into.

**Comment by define-private-public** [comment]
I'll let your pull the trigger to merge.

**Pull request merged**